File: common/src/main/java/cc/turtl/chiselmon/util/format/ComponentUtils.java
package cc.turtl.chiselmon.util.format;

import cc.turtl.chiselmon.ChiselmonConstants;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.MutableComponent;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Iterator;
import java.util.function.Function;

/**
 * Provides basic chat component formatting utils.
 */
public final class ComponentUtils {
    private ComponentUtils() {}

    /** Placeholder component for missing or null data. */
    public static final Component UNKNOWN = literal("???", ColorUtils.DARK_GRAY);

    public static final Component SPACE = Component.literal(" ");

    /**
     * Creates a translatable component prefixed with the mod ID.
     * <p>Example: {@code translatable("msg.caught", "Pikachu")} maps to
     * {@code "chiselmon.msg.caught": "Caught %s!"} in lang files.</p>
     */
    public static MutableComponent modTranslatable(String key, Object... args) {
        return Component.translatable(ChiselmonConstants.MOD_ID + "." + key, args);
    }

    /**
     * Creates a literal component with a specific color.
     * <p>Example: {@code literal("Lvl 50", ColorUtil.GOLD)}</p>
     */
    public static MutableComponent literal(Object text, int color) {
        return Component.literal(text.toString()).withColor(color);
    }

    /**
     * Creates a component from a label and value pair.
     * <p>Example: {@code label("Ability", "Intimidate")} -> "Ability: Intimidate"</p>
     */
    public static Component labelled(@NotNull Object label, @Nullable Object value) {
        MutableComponent labelComp = (label instanceof Component c ? c.copy() : Component.literal(label.toString()));
        // If we pass a component label, it gets overridden with gray for label consistency
        labelComp.withColor(ColorUtils.LIGHT_GRAY);

        Component valueComp = (value == null) ? UNKNOWN :
                (value instanceof Component c ? c : literal(value.toString(), ColorUtils.WHITE));

        return labelComp.append(literal(": ", ColorUtils.LIGHT_GRAY)).append(valueComp);
    }

    /**
     * Joins items into a single component with separators.
     * <p>Example: {@code join(list, ", ", item -> literal(item, ColorUtil.RED))}</p>
     */
    public static <E> Component join(@Nullable Iterable<E> items, String separator, Function<E, Component> mapper) {
        if (items == null || !items.iterator().hasNext()) return UNKNOWN;

        MutableComponent result = Component.empty();
        Iterator<E> it = items.iterator();

        while (it.hasNext()) {
            Component mapped = mapper.apply(it.next());
            if (mapped != null) {
                result.append(mapped);
                if (it.hasNext()) result.append(literal(separator, ColorUtils.LIGHT_GRAY));
            }
        }
        return result;
    }
}

File: common/src/main/kotlin/cc/turtl/chiselmon/core/util/format/ComponentUtils.kt
package cc.turtl.chiselmon.core.util.format

import cc.turtl.turtlshell.api.core.format.ColorLib
import cc.turtl.turtlshell.api.core.util.getRatioColor
import net.minecraft.ChatFormatting
import net.minecraft.network.chat.ClickEvent
import net.minecraft.network.chat.Component
import net.minecraft.network.chat.MutableComponent

/**
 * Placeholder component for missing or null data.
 */
val UNKNOWN: Component = createComponent("???", ColorLib.DARK_GRAY.rgb)
val NONE: Component = createComponent("None", ColorLib.DARK_GRAY.rgb)
val SPACE: Component = Component.literal(" ")
val RESET: Component = Component.literal("").withStyle(ChatFormatting.RESET)

fun createComponent(text: Any?): MutableComponent {
    return createComponent(text, ColorLib.WHITE.rgb)
}

/**
 * Creates a component with a specific color.
 *
 * Example: `literal("Lvl 50", ColorLib.GOLD)`
 */
fun createComponent(text: Any?, color: Int, bold: Boolean = false): MutableComponent {
    val content = text?.toString() ?: ""
    val component = Component.literal(content)

    return component.withStyle { style ->
        style.withColor(color).withBold(bold)
    }
}

/**
 * Creates a clickable URL component that opens the link in the browser when clicked.
 */
fun clickableUrl(url: String): Component {
    return Component.literal(url).withStyle { style ->
        style.withClickEvent(ClickEvent(ClickEvent.Action.OPEN_URL, url))
            .withColor(ColorLib.PINK.rgb)
            .withUnderlined(true)
    }
}

/**
 * Creates a component from a label and value pair.
 *
 * Example: `label("Ability", "Intimidate")` -> "Ability: Intimidate"
 */
fun labelled(label: Any, value: Any?): MutableComponent {
    val labelComp = (label as? Component)?.copy() ?: Component.literal(label.toString())
    val styledLabelComp = labelComp.withStyle { style ->
        style.withColor(ColorLib.LIGHT_GRAY.rgb).withBold(false)
    }

    val valueComp = value?.let {
        when (it) {
            is Component -> it
            else -> createComponent(it.toString(), ColorLib.WHITE.rgb)
        }
    } ?: UNKNOWN

    return styledLabelComp.append(createComponent(": ", ColorLib.LIGHT_GRAY.rgb)).append(valueComp)
}

/**
 * Joins items into a single component with separators.
 *
 * Example: `join(list, ", ") { item -> createComponent(item, ColorLib.RED.rgb) }`
 */
fun <E> join(items: Iterable<E>?, separator: String, mapper: (E) -> Component?): Component {
    val parts = items?.mapNotNull(mapper)?.takeIf { it.isNotEmpty() } ?: return UNKNOWN
    val sep = createComponent(separator, ColorLib.DARK_GRAY.rgb)
    val result = Component.empty()
    parts.forEachIndexed { i, comp ->
        if (i > 0) result.append(sep)
        result.append(comp)
    }
    return result
}

/**
 * Creates a component where the text is colored with a multi-point gradient.
 *
 * @param text The string to color.
 * @param colors The RGB color stops (0xRRGGBB).
 * @return A MutableComponent containing the gradient text.
 */
fun gradient(text: String?, vararg colors: Int): MutableComponent {
    if (text.isNullOrEmpty()) return Component.empty()
    if (colors.isEmpty()) return Component.literal(text)
    if (colors.size == 1) return createComponent(text, colors[0])

    val result = Component.empty()
    val length = text.length

    for (i in text.indices) {
        // Calculate ratio (0.0 to 1.0) based on character index
        val ratio = if (length > 1) i.toFloat() / (length - 1).toFloat() else 0.0f
        val color = getRatioColor(ratio, *colors)
        result.append(Component.literal(text[i].toString()).withStyle { it.withColor(color) })
    }

    return result
}